### PR TITLE
xpra: Add NVENC support

### PR DESCRIPTION
**Pull request by zhaofengli**
###### Motivation for this change

This PR adds a variant of Xpra (`xpraWithNvenc`) with the NVENC encoder which requires proprietary NVIDIA drivers. Tested with the Tesla M40 card with [a workaround patch](https://github.com/Xpra-org/xpra/issues/3136) (not included). Upstream documentations are available [here](https://github.com/Xpra-org/xpra/blob/master/docs/Usage/NVENC.md).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 124335` at f7e92497 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/c7208f0eea39c75d30a4e0b58598dfa0">ib-controller</a></li>
  </ul>
</details>
<details>
  <summary>3 packages built successfully:</summary>
  <ul>
    <li>nv-codec-headers-10</li>
    <li>run-scaled</li>
    <li>xpra</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

**Comment by mvnetbiz** [comment]
Awesome, I have an nvidia card I will test this today with.

File: pkgs/tools/X11/xpra/default.nix [hunk 1]
```diff
@@ -111,8 +121,7 @@ in buildPythonApplication rec {
       --set XPRA_XKB_CONFIG_ROOT "${xorg.xkeyboardconfig}/share/X11/xkb"
       --prefix LD_LIBRARY_PATH : ${libfakeXinerama}/lib
       --prefix PATH : ${lib.makeBinPath [ getopt xorgserver xauth which util-linux pulseaudio ]}
-    )
-  '';
+  '' + (if withNvenc then "--prefix LD_LIBRARY_PATH : ${nvidia_x11}/lib" else "") + ")";
```
Comment (SuperSandro2000) [inline pkgs/tools/X11/xpra/default.nix:124]:
```suggestion
  '' + lib.optionalString withNvenc "--prefix LD_LIBRARY_PATH : ${nvidia_x11}/lib";
```

File: pkgs/tools/X11/xpra/default.nix [hunk 2]
```diff
@@ -81,13 +90,14 @@ in buildPythonApplication rec {
...
-  ];
+  ] ++ lib.optional withNvenc nvencHeaders;
   propagatedBuildInputs = with python3.pkgs; [
     pillow rencode pycrypto cryptography pycups lz4 dbus-python
     netifaces numpy pygobject3 pycairo gst-python pam
     pyopengl paramiko opencv4 python-uinput pyxdg
     ipaddress idna pyinotify
-  ];
+  ]
+  ++ lib.optionals withNvenc (with python3.pkgs; [pynvml pycuda]);
```
Comment (SuperSandro2000) [inline pkgs/tools/X11/xpra/default.nix:100]:
```suggestion
  ] ++ lib.optionals withNvenc (with python3.pkgs; [ pynvml pycuda ]);
```

File: pkgs/tools/X11/xpra/default.nix [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
 , fetchurl
 , fetchpatch
-, substituteAll, python3, pkg-config, writeText
+, substituteAll, python3, pkg-config, runCommand, writeText
 , xorg, gtk3, glib, pango, cairo, gdk-pixbuf, atk, pandoc
 , wrapGAppsHook, xorgserver, getopt, xauth, util-linux, which
 , ffmpeg, x264, libvpx, libwebp, x265, librsvg
 , libfakeXinerama
 , gst_all_1, pulseaudio, gobject-introspection
+, withNvenc ? false, cudatoolkit ? null, nv-codec-headers-10 ? null, nvidia_x11 ? null
```
Comment (SuperSandro2000) [inline pkgs/tools/X11/xpra/default.nix:10]:
```suggestion
, withNvenc ? false, cudatoolkit, nv-codec-headers-10, nvidia_x11
```

**Review by mvnetbiz — APPROVED** [review]
Works for me with 1 window on a gtx 1080. I guess it encodes each window separately and "consumer" cards can only encode one stream at a time without this patch. https://github.com/keylase/nvidia-patch

**Comment by mvnetbiz** [comment]
That applies to the nvidia driver, not nvenc libs, and everything here looks good.

File: pkgs/tools/X11/xpra/default.nix [hunk 4]
```diff
@@ -111,8 +120,7 @@ in buildPythonApplication rec {
       --set XPRA_XKB_CONFIG_ROOT "${xorg.xkeyboardconfig}/share/X11/xkb"
       --prefix LD_LIBRARY_PATH : ${libfakeXinerama}/lib
       --prefix PATH : ${lib.makeBinPath [ getopt xorgserver xauth which util-linux pulseaudio ]}
-    )
-  '';
+  '' + lib.optionalString withNvenc "--prefix LD_LIBRARY_PATH : ${nvidia_x11}/lib" + ")";
```
Comment (veprbl) [inline pkgs/tools/X11/xpra/default.nix:123]:
```suggestion
  '' + lib.optionalString withNvenc ''
      --prefix LD_LIBRARY_PATH : ${nvidia_x11}/lib
  '' + ''
    )
  '';
```
This should make it a non-rebuld

**Comment by veprbl** [comment]
Result of `nixpkgs-review pr 124335` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>nv-codec-headers-10</li>
  </ul>
</details>

**Pull request merged**